src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -2944,6 +2952,30 @@ private LLVMValueRef GetFieldAddress(FieldDesc field, bool isStatic)
         /// </summary>
         private void TriggerCctor(MetadataType type)
         {
+            ISymbolNode classConstructionContextSymbol = _compilation.NodeFactory.TypeNonGCStaticsSymbol(type);
+            _dependencies.Add(classConstructionContextSymbol);
+            LLVMValueRef firstNonGcStatic = LoadAddressOfSymbolNode(classConstructionContextSymbol);
+
+            // TODO: Codegen could check whether it has already run rather than calling into EnsureClassConstructorRun
+            // but we'd have to figure out how to manage the additional basic blocks
+            LLVMValueRef classConstructionContextPtr = LLVM.BuildGEP(_builder, firstNonGcStatic, new LLVMValueRef[] { BuildConstInt32(-2) }, "classConstructionContext");
+            StackEntry classConstructionContext = new AddressExpressionEntry(StackValueKind.NativeInt, "classConstructionContext", classConstructionContextPtr, GetWellKnownType(WellKnownType.IntPtr));
+            CallRuntime("System.Runtime.CompilerServices", _compilation.TypeSystemContext, ClassConstructorRunner, "EnsureClassConstructorRun", new StackEntry[] { classConstructionContext });
+        }
+
+        /// <summary>
+        /// Triggers creation of thread static storage and the static constructor if present
+        /// </summary>
+        private ISymbolNode TriggerCctorWithThreadStaticStorage(MetadataType type, bool needsCctorCheck, out ExpressionEntry returnExp)
+        {
+            ISymbolNode threadStaticIndexSymbol = _compilation.NodeFactory.TypeThreadStaticIndex(type);
+            LLVMValueRef threadStaticIndex = LoadAddressOfSymbolNode(threadStaticIndexSymbol);
+
+            StackEntry typeManagerSlotEntry = new LoadExpressionEntry(StackValueKind.ValueType, "typeManagerSlot", threadStaticIndex, GetWellKnownType(WellKnownType.Int32));
+            LLVMValueRef typeTlsIndexPtr =
+                LLVM.BuildGEP(_builder, threadStaticIndex, new LLVMValueRef[] { BuildConstInt32(1) }, "typeTlsIndexPtr"); // index is the second field after the ptr.
+            StackEntry tlsIndexExpressionEntry = new LoadExpressionEntry(StackValueKind.ValueType, "typeTlsIndex", typeTlsIndexPtr, GetWellKnownType(WellKnownType.Int32));
+
             if (_compilation.TypeSystemContext.HasLazyStaticConstructor(type))
```

Comment:
There's something I dont understand about `(owningType.IsBeforeFieldInit || (!owningType.IsBeforeFieldInit && owningType != _thisType))` which must mean it is not causing this bool to be false in any of the tests paths, as this change passes the tests. Addressed in latest push.

Reply:
> There's something I dont understand

It should cover cases like this:

```csharp
class Foo
{
    [ThreadStatic]
    static int x;

    static Foo() { x = 123 }

    static void Do()
    {
        // No cctor check needed before accessing x because
        // we had to trigger the cctor of Foo to call Do()
        x = 456;
    }
}
```

Reply:
I see, `IsBeforeFieldInit` would be false in the `Do` field getter, thanks.